sky/serve/replica_managers.py

```diff
@@ -0,0 +1,804 @@
+                    info.status_property.sky_down_status = (
+                        ProcessStatus.SUCCEEDED)
+                # Failed replica still count as a replica. In our current
+                # design, we want to fail early if user code have any error.
+                # This will prevent infinite loop of teardown and
+                # re-provision. However, there is a special case that if the
+                # replica is UP for longer than initial_delay_seconds, we
+                # assume it is just some random failure and we should restart
+                # the replica. Please refer to the implementation of
+                # `is_scale_down_succeeded` for more details.
+                # TODO(tian): Currently, restart replicas that failed within
+                # initial_delay_seconds is not supported. We should add it
+                # later when we support `sky serve update`.
+                if info.status_property.is_scale_down_succeeded(
+                        self.initial_delay_seconds, self.auto_restart):
+                    # This means the cluster is deleted due to
+                    # a scale down or the cluster is recovering
+                    # from preemption. Delete the replica info
+                    # so it won't count as a replica.
+                    serve_state.remove_replica(self.service_name, replica_id)
+                    if info.status_property.preempted:
+                        removal_reason = 'for preemption recovery'
+                    else:
+                        removal_reason = 'normally'
+                    logger.info(f'Replica {replica_id} removed from the '
+                                f'replica table {removal_reason}.')
+                else:
+                    logger.info(f'Termination of replica {replica_id} '
+                                'finished. Replica info is kept since some '
+                                'failure detected.')
+                    serve_state.add_or_update_replica(self.service_name,
+                                                      replica_id, info)
+
+    def _process_pool_refresher(self) -> None:
+        """Periodically refresh the launch/down process pool."""
+        while True:
+            logger.debug('Refreshing process pool.')
+            try:
+                self._refresh_process_pool()
+            except Exception as e:  # pylint: disable=broad-except
+                # No matter what error happens, we should keep the
+                # process pool refresher running.
+                logger.error(f'Error in process pool refresher: {e}')
+            time.sleep(_PROCESS_POOL_REFRESH_INTERVAL)
+
+    @with_lock
+    def _fetch_job_status(self) -> None:
+        """Fetch the service job status of all replicas.
+
+        This function will monitor the job status of all replicas
+        to make sure the service is running correctly. If any of the
+        replicas failed, it will terminate the replica.
```

Comment:
Now added:
```
It is still needed even if we already keep probing the replicas,
since the replica job might launch the API server in the background
(using &), and the readiness probe will not detect the worker failure.
```

Reply:
Sorry, I don't understand the explanation here. The readiness probe will fail to connect to the service in the case you mentioned here, and the job_status will be `SUCCEEDED` as the job submitted to the worker is finished. How does the job status help the case?

Reply:
Suppose we are using this YAML:

```yaml
run: |
  python3 api_server.py & # not blocking
  python3 worker.py # blocking
```

If the `worker.py ` fails but the `api_server.py` is still working, the probe will still succeed (suppose we are not using a real workload as the probe), and the job status fails.

Reply:
This could happen when the probe is misconfigured, or probably you are using sth like:

```yaml
run: |
  python3 api_server.py & # not blocking
  python3 worker.py & # not blocking
  python3 some_ auxiliary_process.py # blocking
```

and the `some_daemon_process.py` failed, but the `worker.py` is still working (for a short time and will eventually output the wrong answer due to missing of auxiliary process).